Do not give function allocations alignment in consteval and Miri.

zachs18 opened the pull request:
We do not yet have a (clear and T-lang approved) design for how `#[align(N)]` on functions should affect function pointers' addresses on various platforms, so for now do not give function pointers alignment in consteval and Miri.

----

Old summary:

Not a full solution to <https://github.com/rust-lang/rust/issues/144661>, but fixes the immediate issue by making function allocations all have alignment 1 in consteval, ignoring `#[rustc_align(N)]`, so the compiler doesn't know if any offset other than 0 is non-null.

A more "principlied" solution would probably be to make function pointers to `#[instruction_set(arm::t32)]` functions be at offset 1 of an align-`max(2, align attribute)` allocation instead of at offset 0 of their allocation during consteval, and on wasm to either disallow `#[align(N)]` where N > 1, or to pad the function table such that the function pointer of a `#[align(N)]` function is a multiple of `N` at runtime.

<!-- homu-ignore:start -->
<!--
If this PR is related to an unstable feature or an otherwise tracked effort,
please link to the relevant tracking issue here. If you don't know of a related
tracking issue or there are none, feel free to ignore this.

This PR will get automatically assigned to a reviewer. In case you would like
a specific user to review your work, you can assign it to them by using

    r? <reviewer name>
-->
<!-- homu-ignore:end -->

zachs18 committed:
Do not give function allocations alignment in consteval or miri.

rust-log-analyzer commented:
The job **`x86_64-gnu-miri`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/16634694052/job/47072712300) [(plain enhanced)](https://triage.rust-lang.org/gha-logs/rust-lang/rust/47072712300) [(plain)](https://github.com/rust-lang/rust/commit/67f6fa1e6bb2ba65a955e385741a1c25201f0d6f/checks/47072712300/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
tests/pass/shims/x86/intrinsics-x86-avx2.rs ... ok
tests/pass/shims/x86/intrinsics-x86-gfni.rs ... ok

FAILED TEST: tests/pass/fn_align.rs
command: MIRI_ENV_VAR_TEST="0" MIRI_TEMP="/tmp/miri-uitest-x6BrGt" RUST_BACKTRACE="1" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/bin/miri" "--error-format=json" "--sysroot=/checkout/obj/build/x86_64-unknown-linux-gnu/miri-sysroot" "-Dwarnings" "-Dunused" "-Ainternal_features" "-Zui-testing" "--out-dir" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage1-tools/x86_64-unknown-linux-gnu/tmp/miri_ui/0/tests/pass" "tests/pass/fn_align.rs" "-Zmin-function-alignment=8" "--edition" "2021"

error: test got exit status: 101, but expected 0
 = note: the compiler panicked

error: no output was expected
Execute `./miri test --bless` to update `tests/pass/fn_align.stderr` to the actual output
+++ <stderr output>

thread 'main' panicked at tests/pass/fn_align.rs:20:5:
assertion failed: (foo as usize).is_multiple_of(256)
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
note: in Miri, you may have to set `MIRIFLAGS=-Zmiri-env-forward=RUST_BACKTRACE` for the environment variable to have an effect


full stderr:

thread 'main' panicked at tests/pass/fn_align.rs:20:5:
assertion failed: (foo as usize).is_multiple_of(256)
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
note: in Miri, you may have to set `MIRIFLAGS=-Zmiri-env-forward=RUST_BACKTRACE` for the environment variable to have an effect

full stdout:


FAILURES:
---

Location:
   /cargo/registry/src/index.crates.io-1949cf8c6b5b557f/ui_test-0.30.2/src/lib.rs:365

Backtrace omitted. Run with RUST_BACKTRACE=1 environment variable to display it.
Run with RUST_BACKTRACE=full to include source snippets.
error: test failed, to rerun pass `--test ui`

Caused by:
  process didn't exit successfully: `/checkout/obj/build/x86_64-unknown-linux-gnu/stage1-tools/x86_64-unknown-linux-gnu/release/deps/ui-22b922dd454f6d1c` (exit status: 1)
Command has failed. Rerun with -v to see more details.
Build completed unsuccessfully in 0:39:57
  local time: Wed Jul 30 22:48:16 UTC 2025
  network time: Wed, 30 Jul 2025 22:48:16 GMT
##[error]Process completed with exit code 1.
Post job cleanup.
```

</details>

RalfJung commented:
So this essentially reverts https://github.com/rust-lang/rust/pull/140072... Cc @folkertdev.

But I agree, https://github.com/rust-lang/rust/pull/140072 is just wrong, so we should probably revert and reconsider.

RalfJung commented:
@zachs18 you can just remove that failing Miri test.

folkertdev commented:
For wasm we already basically decided to limit the maximum alignment to just 1 (the only value that makes sense on that target). It's unfortunate because then any code that uses an alignment higher than 1 is non-portable, but function alignment is a highly specific tool anyway so that should be OK.

> make function pointers to `#[instruction_set(arm::t32)]` functions be at offset 1

I guess that works? Seems a bit hacky but at least const eval won't need to actually know about the pointer tagging.

RalfJung commented:
That's just the one case where we know about this kind of problem though. We should have a very high bar for such architecture-specific logic in the interpreter and I am not convinced this meets that bar.

RalfJung commented:
Given that this turns out to be a non-trivial question, I think that until we have a clear motivation for turning `align(N)` on functions into a guarantee about function pointers, *and* a t-lang decision to do so, we shouldn't use this attribute at all in the interpreter.

zachs18 commented:
@rustbot ready

I deleted the Miri test, and changed the FIXME comment from "as a workaround for #144661, [...]" to "Until we have a clear design for the effects of align(N) functions on the address of function pointers, we don't consider the align(N) attribute on functions in the interpreter."

RalfJung commented on compiler/rustc_const_eval/src/interpret/memory.rs:
```diff
@@ -937,8 +937,12 @@ impl<'tcx, M: Machine<'tcx>> InterpCx<'tcx, M> {
+                FnVal::Instance(_instance) => {
+                    // FIXME: Until we have a clear design for the effects of align(N) functions
+                    // on the address of function pointers, we don't consider the align(N)
+                    // attribute on functions in the interpreter.
+                    //self.tcx.codegen_instance_attrs(instance.def).alignment.unwrap_or(Align::ONE)
```
```suggestion
                    // attribute on functions in the interpreter.
```
Code in comments just rots, let's get rid of it.

nikic commented:
@RalfJung I'm referring to:

> `F<type><abi>`
>
> This specifies the alignment for function pointers. The options for `<type>` are:
> * i: The alignment of function pointers is independent of the alignment of functions, and is a multiple of `<abi>`.
> * n: The alignment of function pointers is a multiple of the explicit alignment specified on the function, and is a multiple of `<abi>`.

The ARM targets specify `Fi8` in the data layout, which means they have 1 byte function pointer alignment that is independent of function alignment.

zachs18 committed:
Update compiler/rustc_const_eval/src/interpret/memory.rs

Replace commented-out code with link to context for change.

Co-authored-by: Ralf Jung <post@ralfj.de>

RalfJung commented:
Thanks!

@bors r+

bors commented:
:pushpin: Commit fe720181b52157f9a7195b3d9b0b28b20dc9583d has been approved by `RalfJung`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=RalfJung fe720181b52157f9a7195b3d9b0b28b20dc9583d -->
<!-- homu: {"type":"Approved","sha":"fe720181b52157f9a7195b3d9b0b28b20dc9583d","approver":"RalfJung","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->
